File: typescript/default-general/tools/darena-solutions/PatientAgeTool.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp";
import { Request } from "express";
import { IMcpTool } from "../../IMcpTool";
import { z } from "zod";
import {
  getFhirContext,
  getPatientIdIfContextExists,
} from "../../fhir-utilities";
import { createTextResponse } from "../../mcp-utilities";
import axios from "axios";
import { FhirContext } from "../../fhir-context";
import { differenceInYears, parseISO } from "date-fns";

type Patient = {
  id: string;
  birthDate?: string;
};

class PatientAgeTool implements IMcpTool {
  registerTool(server: McpServer, req: Request) {
    server.tool(
      "GetPatientAge",
      "Gets the age of a patient. If patient context already exists, then the first name and last name is not required.",
      {
        firstName: z.string().optional().describe("The patient's first name"),
        lastName: z.string().optional().describe("The patient's last name"),
      },
      async ({ firstName, lastName }) => {
        const fhirContext = getFhirContext(req);
        if (!fhirContext) {
          return createTextResponse(
            "A FHIR server url or token was not provided in the http context.",
            { isError: true },
          );
        }

        const patientIdContext = getPatientIdIfContextExists(req);
        let patient: Patient | null = null;
        if (patientIdContext) {
          try {
            const response = await axios.get(
              `${fhirContext.url}/Patient/${patientIdContext}`,
              { headers: { Authorization: `Bearer ${fhirContext.token}` } },
            );

            patient = response.data;
          } catch {
            return createTextResponse(
              "Could not retrieve the patient from the context.",
            );
          }
        } else {
          if (!firstName && !lastName) {
            return createTextResponse(
              "Patient context was not found, thus the first name, last name, or both is required.",
              { isError: true },
            );
          }

          let patients = await this._patientSearcher(
            fhirContext,
            firstName,
            lastName,
          );

          if (!patients?.length) {
            patients = await this._patientSearcher(
              fhirContext,
              lastName,
              firstName,
            );
          }

          if (patients && patients.length > 1) {
            return createTextResponse(
              "More than one patient was found. Provide more details.",
              { isError: true },
            );
          }

          patient = patients?.length ? patients[0] : null;
        }

        if (!patient) {
          return createTextResponse("The patient could not be found.", {
            isError: true,
          });
        }

        if (!patient.birthDate) {
          return createTextResponse(
            "A birth date could not be found for the patient.",
            { isError: true },
          );
        }

        try {
          const date = parseISO(patient.birthDate);
          const age = differenceInYears(new Date(), date);

          return createTextResponse(`The patient's age is: ${age}`);
        } catch {
          return createTextResponse(
            "Could not parse the patient's birth date.",
            { isError: true },
          );
        }
      },
    );
  }

  private async _patientSearcher(
    fhirContext: FhirContext,
    searchFirstName: string | null | undefined,
    searchLastName: string | null | undefined,
  ): Promise<Patient[] | null> {
    const searchParameters: string[] = [];
    if (searchFirstName) {
      searchParameters.push(`given=${searchFirstName}`);
    }

    if (searchLastName) {
      searchParameters.push(`family=${searchLastName}`);
    }

    const url = `${fhirContext.url}/Patient?${searchParameters.join("&")}`;
    const { data } = await axios.get(url, {
      headers: {
        Authorization: `Bearer ${fhirContext.token}`,
      },
    });

    return data?.entry.length
      ? (data.entry as { resource: Patient }[]).map((x) => x.resource)
      : null;
  }
}

export const PatientAgeToolInstance = new PatientAgeTool();

File: dotnet/default/DarenaHealth.Community.McpTools/DarenaSolutions/PatientAgeTool.cs
using DarenaHealth.Community.Mcp.Core;
using DarenaHealth.Community.Mcp.Core.Extensions;
using DarenaHealth.Community.Mcp.Core.Models;
using Hl7.Fhir.Model;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.DependencyInjection;
using ModelContextProtocol.Protocol;
using ModelContextProtocol.Server;

namespace DarenaHealth.Community.McpTools.DarenaSolutions;

public class PatientAgeTool : IMcpTool
{
    private const string FirstNameParameter = "firstName";
    private const string LastNameParameter = "lastName";

    public string Name { get; } = "GetPatientAge";

    public string? Description { get; } =
        "Gets the age of a patient. If patient context already exists, then the first name and last name is not required.";

    public List<McpToolArgument> Arguments { get; } =
        [
            new McpToolArgument
            {
                Type = "string",
                Name = FirstNameParameter,
                Description = "The patient's first name",
                IsRequired = false,
            },
            new McpToolArgument
            {
                Type = "string",
                Name = LastNameParameter,
                Description = "The patient's last name",
                IsRequired = false,
            },
        ];

    public List<string> FhirScopes { get; } = ["patient/Patient.read"];

    public async Task<CallToolResponse> HandleAsync(
        HttpContext httpContext,
        IMcpServer mcpServer,
        CallToolRequestParams context
    )
    {
        var fhirContext = httpContext.GetFhirContext();
        if (fhirContext is null)
        {
            return McpToolUtilities.CreateTextToolResponse(
                "A FHIR server url or token was not provided in the http context.",
                isError: true
            );
        }

        if (mcpServer.Services is null)
        {
            return McpToolUtilities.CreateTextToolResponse(
                "An unexpected server error occurred. Services were not found.",
                isError: true
            );
        }

        var searchService = mcpServer.Services.GetRequiredService<IPatientSearchService>();

        Patient? patient = null;
        var patientIdContext = httpContext.GetPatientIdIfContextExists();
        if (!string.IsNullOrWhiteSpace(patientIdContext))
        {
            patient = await searchService.FindByIdAsync(fhirContext, patientIdContext);
        }
        else
        {
            if (context.Arguments is null)
            {
                return McpToolUtilities.CreateTextToolResponse(
                    "No arguments were specified",
                    isError: true
                );
            }

            var firstName = context.GetArgumentValueOrNull(FirstNameParameter);
            var lastName = context.GetArgumentValueOrNull(LastNameParameter);

            if (string.IsNullOrWhiteSpace(firstName) && string.IsNullOrWhiteSpace(lastName))
            {
                return McpToolUtilities.CreateTextToolResponse(
                    "Patient context was not found, thus the first name, last name, or both is required.",
                    isError: true
                );
            }

            var patients = await searchService.SearchByNameAsync(fhirContext, firstName, lastName);
            if (patients.Count == 0)
            {
                patients = await searchService.SearchByNameAsync(fhirContext, lastName, firstName);
            }

            if (patients.Count > 1)
            {
                return McpToolUtilities.CreateTextToolResponse(
                    "More than one patient found. Provide additional details.",
                    isError: true
                );
            }

            if (patients.Count == 1)
            {
                patient = patients[0];
            }
        }

        if (patient is null)
        {
            return McpToolUtilities.CreateTextToolResponse(
                "The patient could not be found.",
                isError: true
            );
        }

        var patientDob = patient.BirthDateElement?.ToDate()?.ToDateTime();
        if (patientDob?.Years is null)
        {
            return McpToolUtilities.CreateTextToolResponse(
                "This patient does not have a birth date set or no year information is present.",
                isError: true
            );
        }

        var dobConverted = new DateTime(
            patientDob.Years.Value,
            patientDob.Months ?? 1,
            patientDob.Days ?? 1
        );
        var now = DateTime.UtcNow;
        var age = now.Year - dobConverted.Year;

        if (dobConverted > now.AddYears(-age))
        {
            age--;
        }

        return McpToolUtilities.CreateTextToolResponse($"The patient's age is: {age}");
    }
}
